packages/app/src/providers/api.ts

```diff
@@ -55,26 +72,41 @@ export abstract class ProviderApi {
   protected async cancellableFetchValue<T>(
     endpoint: string,
     storeParams: ValuesStoreParams,
-    queryParams?: Record<string, string | boolean | undefined>,
+    queryParams: Record<string, string | boolean | undefined>,
     responseType?: ResponseType
   ): Promise<AxiosResponse<T>> {
     const cancelSource = axios.CancelToken.source();
     const request = { storeParams, cancelSource };
     this.valueRequests.add(request);
 
+    this.progress.set(storeParams, 0);
+    this.notifyProgressChange();
+
     try {
       const { token: cancelToken } = cancelSource;
       return await this.client.get<T>(endpoint, {
         cancelToken,
-        params: queryParams || storeParams,
+        params: queryParams,
         responseType,
+        onDownloadProgress: (evt: ProgressEvent) => {
+          if (evt.lengthComputable) {
+            this.progress.set(storeParams, evt.loaded / evt.total);
+            this.notifyProgressChange();
+          }
+        },
       });
     } finally {
-      // Remove cancellation source when request fulfills
+      // Remove cancellation source and progress when request fulfills
       this.valueRequests.delete(request);
+      this.progress.delete(storeParams);
+      this.notifyProgressChange();
     }
   }
 
+  private notifyProgressChange() {
+    this.progressListeners.forEach((cb) => cb([...this.progress.values()]));
```

Comment:
Just a thought, if `progressListeners` was a `Map` instead of a `Set`, you could notify only the relevant change, right ?

You would need to pass the `storeParams` as arg of most callbacks though.

Reply:
You mean so listeners only get notified of the request that has "progressed"? Not sure there's a need for this. The value loader loops through all the progress values on every render anyway. We don't have a component that would benefit from re-rendering only when a specific request's progress changes.